File: Diagrammatize.Design.Pattern/code/cxx/Flyweight/Sample/BigCharFactory.cpp
#include "BigCharFactory.hpp"

BigCharFactory::BigCharFactory() {
}

BigCharFactory& BigCharFactory::getInstance() {
    static BigCharFactory singleton;
    return singleton;
}

BigChar* BigCharFactory::getBigChar(char charname) {
    std::lock_guard<std::mutex> lck(pool_mutex);
    BigChar* bc = pool[std::string("") + charname];
    if (bc == nullptr) {
        bc = new BigChar(charname);
        pool[std::string("") + charname] = bc;
    }
    return bc;
}

File: Diagrammatize.Design.Pattern/code/python/Flyweight/A1/BigCharFactory.py

import threading
from BigChar import BigChar

class BigCharFactory:
    singleton = None

    def __init__(self):
        self.pool = dict()
        self.lock = threading.Lock()

    @classmethod
    def getInstance(cls):
        if not cls.singleton:
            cls.singleton = cls()
        return cls.singleton

    def getBigChar(self, charname):
        with self.lock:
            bc = self.pool.get(charname)
            if bc is None:
                bc = BigChar(charname)
                self.pool[charname] = bc
            return bc



